sdk/storage/azure_storage_blob/src/streams/partitioned_stream.rs

```diff
@@ -0,0 +1,187 @@
+use std::{
+    mem,
+    pin::{pin, Pin},
+    task::Poll,
+};
+
+use azure_core::stream::SeekableStream;
+use bytes::Bytes;
+use futures::{ready, stream::FusedStream, AsyncRead, Stream};
+
+type AzureResult<T> = azure_core::Result<T>;
+
+pub(crate) struct PartitionedStream {
+    inner: Box<dyn SeekableStream>,
+    buf: Vec<u8>,
+    partition_len: usize,
+    buf_offset: usize,
+    total_read: usize,
+    inner_complete: bool,
+}
+
+impl PartitionedStream {
+    pub(crate) fn new(inner: Box<dyn SeekableStream>, partition_len: usize) -> Self {
+        assert!(partition_len > 0);
+        Self {
+            buf: vec![0u8; std::cmp::min(partition_len, inner.len())],
+            inner,
+            partition_len,
+            buf_offset: 0,
+            total_read: 0,
+            inner_complete: false,
+        }
+    }
+
+    fn take(&mut self) -> Vec<u8> {
+        let mut ret = mem::replace(
+            &mut self.buf,
+            vec![0u8; std::cmp::min(self.partition_len, self.inner.len() - self.total_read)],
+        );
+        ret.truncate(self.buf_offset);
+        self.buf_offset = 0;
+        ret
+    }
+}
+
+impl Stream for PartitionedStream {
+    type Item = AzureResult<Bytes>;
+
+    fn poll_next(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        loop {
+            if this.inner_complete || this.buf_offset >= this.buf.len() {
```

Comment:
I am wondering if the RHS expression should instead be:
```rust
this.buf_offset == this.buf.len()
```
If I am understanding correctly, wouldn't the case that the buffer offset being greater than the buffer size mean we will lose that data:
`(this.buf_offset) - (buffer size) = # of bytes being lost`

Since we did in fact read it out of the stream (as evidenced by `this.buf_offset`) but it's bigger than the buffer, so those extra bytes got eaten up by the aether?

I think this is a super super edge case since we really shouldn't ever get into that case, but wondering if we should move the "greater-than" case to raise an error.